library/core/src/ptr/mod.rs

```diff
@@ -63,12 +63,212 @@
 //! separate allocated object), heap allocations (each allocation created by the global allocator is
 //! a separate allocated object), and `static` variables.
 //!
+//!
+//! # Strict Provenance
+//!
+//! This section is *non-normative* and describes an experimental set of APIs that help tools
+//! that validate the memory-safety of your program's execution. Notably this includes [miri][]
+//! and [CHERI][], which can detect when you access out of bounds memory or otherwise violate
+//! Rust's memory model.
+//!
+//! **The following text is overly strict and insufficiently formal, and is an extremely
+//! strict interpretation of provenance.** Provenance must exist in some form for any language
+//! higher-level than an assembler, but to our knowledge no one has ever been able to square
+//! the notion of provenance with all the operations that programmers believe should be permitted.
```

Comment:
That's rather vague (arguably C's PNVI proposals at least get very close). I'd just avoid statements like this and instead say that specifying provenance is an ongoing challenge in all low-level languages (insert link to C's efforts).

Reply:
`$ cargo clippy -- -Wclippy::pedantic`
`**warning: babbling commentary ahead**`
` --> library/core/ptr/mod.rs:76:5`
`  |`
`  | //! higher-level than an assembler, but to our knowledge no one has ever been able to square`
`  | ⠀⠀⠀^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^`
Arguably something resembling "provenance" does exist in assembler for CHERI, and also arguably in any situation where violating any other memory protection model could cause program-invalidating behavior based on whether the existing context (either in the pointer or e.g. in the virtual process environment) includes "permission" to access a given memory location. Only in machines specified to have unprotected memory is it strictly permitted to ignore the notion of provenance. Most processes simply respect the implicit rules and the CPU and the OS cooperate to paper over the difference... or they segfault a lot.
`$ █`

Reply:
The key phrase here is "with all the operations that programmers believe should be permitted". Programmers want *so much* random shit to work, and I think even the most conservative C models still slap a bunch of "yeah this is UB" stickers on things people Want To Do.

In terms of *strict* provenance we have many examples: remerging two contiguous slices, xor lists, certain 'downcasting' patterns, reading pointers from disk, etc etc etc.

Reply:
"Provenance must exist in some form for any programming language compiled for modern computer architectures, but specifying a model for provenance in a way that is useful to both compilers and programmers is an ongoing challenge."